pkg/release/build.go

```diff
@@ -219,6 +219,21 @@ func MakeRelease(platform Platform, opts BuildOptions, pkgDir string) error {
 			return errors.Wrapf(err, "%s %s: %s", cmd.Env, cmd.Args, string(stdoutBytes))
 		}
 
+		if opts.BuildTag != "" {
```

Comment:
This logic does not hold after #101943 because we are using this to override the baked-in version for nightlies. Right now, this value being empty means "use the checked-in value from `version.txt`", and the value being set means "override that value with this other value". So in fact this logic will never actually be used for release binaries which is when you want to apply it.

One option is:

* Add an extra configuration like `VerifyVersion bool` to `BuildOptions` which tells this code to additionally sanity-check the build tag vs. the `git describe` output.
* If `VerifyVersion` is set, call `git describe --tags --dirty --match=v[0-9]*` here, then cross-check the output against the output of `cockroach version`.
* Update `pkg/cmd/publish-provisional-artifacts` to add a command-line option like `--verify-version` which sets `VerifyVersion`.
* Update `build/teamcity/internal/release/process/build-cockroach-release.sh` to pass `--verify-version`.

Reply:
I think I had a slightly different goal here -- not so much that I was verifying the `cockroach version` output against the `git describe` output but rather than I'm verifying it against the string that we'll use as the artifact name when we publish said artifact, i.e. if you download/docker pull/k8s deploy "cockroach-X" from the cloud bucket or registry, then that binary should identify itself _as_ "X".

I'm apathetic on how we do this, but do think we should have a test somewhere, before we upload/push the binary, that the name we're uploading/pushing it to is what `version` returns so we don't accidentally break that invariant again?

Reply:
I'm also more than happy to close this and file an issue instead if you think it'd be easier to just do it vs telling me how to do it here -- this stuff has all moved around a bit since I was last near it.

Reply:
filed as https://github.com/cockroachdb/cockroach/issues/102019

Reply:
The new issue LGTM, thanks.